File: 12-rainRisk/python/partTwo.py
from typing import Tuple

from common import *


def make_positive(n: int) -> int:
    return n if n >= 0 else n + -2 * n


def make_negative(n: int) -> int:
    return n if n < 0 else n + -2 * n


def rotate_waypoint(
    current: Tuple[int, int], direction: str, amount: int
) -> Tuple[int, int]:

    assert direction in ["l", "r"], f"invalid rotate direction '{direction}'"

    lat_delta, long_delta = current

    times = int(amount / 90)  # number of times to rotate the waypoint by

    # Determine the current quadrant
    quadrant = None
    if lat_delta >= 0 and long_delta >= 0:
        quadrant = 1
    elif lat_delta < 0 and long_delta >= 0:
        quadrant = 2
    elif lat_delta < 0 and long_delta < 0:
        quadrant = 3
    elif lat_delta >= 0 and long_delta < 0:
        quadrant = 4

    assert quadrant is not None, f"unable to determine quadrant for {current}"

    # Determine the new quadrant
    new_quadrant = quadrant
    if direction == "r":
        new_quadrant += times
    elif direction == "l":
        new_quadrant -= times

    # Loop quadrant around if it went outside of 1-4
    if new_quadrant > 4:
        new_quadrant -= 4
    elif new_quadrant < 1:
        new_quadrant += 4

    # Swap values if the new quadrant is not the opposite one
    quadrant_diff = quadrant - new_quadrant
    if quadrant_diff % 2 != 0:
        t = lat_delta
        lat_delta = long_delta
        long_delta = t

    # Transform deltas into their new quadrant
    if new_quadrant == 1:
        lat_delta = make_positive(lat_delta)
        long_delta = make_positive(long_delta)
    elif new_quadrant == 2:
        lat_delta = make_negative(lat_delta)
        long_delta = make_positive(long_delta)
    elif new_quadrant == 3:
        lat_delta = make_negative(lat_delta)
        long_delta = make_negative(long_delta)
    elif new_quadrant == 4:
        lat_delta = make_positive(lat_delta)
        long_delta = make_negative(long_delta)

    return lat_delta, long_delta


def move_to_waypoint(waypoint_delta: Tuple[int, int], times: int) -> Tuple[int, int]:
    lat_delta, long_delta = waypoint_delta
    return lat_delta * times, long_delta * times


def translate_movement(
    waypoint_delta: Tuple[int, int], instruction: Instruction
) -> Tuple[Tuple[int, int], int, int]:
    # Returns the new current direction and the lat/long delta

    lat_delta = 0
    long_delta = 0

    if instruction.action in ["l", "r"]:
        waypoint_delta = rotate_waypoint(
            waypoint_delta, instruction.action, instruction.magnitude
        )
    elif instruction.action == "f":
        lat_delta, long_delta = move_to_waypoint(waypoint_delta, instruction.magnitude)
    elif instruction.action in ["n", "s", "e", "w"]:
        wp_lat_delta, wp_long_delta = calculate_direction_deltas(
            instruction.action, instruction.magnitude
        )
        wp_lat, wp_long = waypoint_delta
        waypoint_delta = (wp_lat + wp_lat_delta, wp_long + wp_long_delta)
    else:
        raise AssertionError(f"invalid action '{instruction.action}'")

    return waypoint_delta, lat_delta, long_delta


def partTwo(instr: str) -> int:
    input_list = parse(instr)

    waypoint_delta = (1, 10)
    lat = 0  # north/south, pos/neg
    long = 0  # west/east, pos/neg

    for instruction in input_list:
        waypoint_delta, lad, lod = translate_movement(waypoint_delta, instruction)
        lat += lad
        long += lod

    # Calculate manhattan distance
    # If either value is negative, make it positive
    if lat < 0:
        lat = lat + -2 * lat
    if long < 0:
        long = long + -2 * long

    return lat + long

File: 12-rainRisk/go/challenge/partTwo.go
package challenge

import "fmt"

func makePositive(n int) int {
	if n >= 0 {
		return n
	}
	return n + -2*n
}

func makeNegative(n int) int {
	if n < 0 {
		return n
	}
	return n + -2*n
}

func rotateWaypoint(current [2]int, direction string, amount int) [2]int {
	if !(direction == "l" || direction == "r") {
		panic(fmt.Errorf("invalid rotate direction '%s'", direction))
	}

	latDelta := current[0]
	longDelta := current[1]

	times := int(amount / 90)

	var quadrant int
	if latDelta >= 0 && longDelta >= 0 {
		quadrant = 1
	} else if latDelta < 0 && longDelta >= 0 {
		quadrant = 2
	} else if latDelta < 0 && longDelta < 0 {
		quadrant = 3
	} else if latDelta >= 0 && longDelta < 0 {
		quadrant = 4
	}

	if quadrant == 0 {
		panic(fmt.Errorf("unable to determine quadrant for %d", current))
	}

	newQuadrant := quadrant
	if direction == "r" {
		newQuadrant += times
	} else if direction == "l" {
		newQuadrant -= times
	}

	if newQuadrant > 4 {
		newQuadrant -= 4
	} else if newQuadrant < 1 {
		newQuadrant += 4
	}

	quadrantDiff := quadrant - newQuadrant
	if quadrantDiff%2 != 0 {
		t := latDelta
		latDelta = longDelta
		longDelta = t
	}

	if newQuadrant == 1 {
		latDelta = makePositive(latDelta)
		longDelta = makePositive(longDelta)
	} else if newQuadrant == 2 {
		latDelta = makeNegative(latDelta)
		longDelta = makePositive(longDelta)
	} else if newQuadrant == 3 {
		latDelta = makeNegative(latDelta)
		longDelta = makeNegative(longDelta)
	} else if newQuadrant == 4 {
		latDelta = makePositive(latDelta)
		longDelta = makeNegative(longDelta)
	}

	return [2]int{latDelta, longDelta}

}

func moveToWaypoint(waypointDelta [2]int, times int) (int, int) {
	return waypointDelta[0] * times, waypointDelta[1] * times
}

func translateMovementTwo(waypointDelta [2]int, instruction Instruction) ([2]int, int, int) {
	// Returns the new current direction and the lat/long delta

	var (
		latDelta  int
		longDelta int
	)

	if instruction.Action == "l" || instruction.Action == "r" {
		waypointDelta = rotateWaypoint(waypointDelta, instruction.Action, instruction.Magnitude)
	} else if instruction.Action == "f" {
		latDelta, longDelta = moveToWaypoint(waypointDelta, instruction.Magnitude)
	} else if instruction.Action == "n" || instruction.Action == "s" || instruction.Action == "e" || instruction.Action == "w" {
		wpLatDelta, wpLongDelta := calculateDirectionDeltas(instruction.Action, instruction.Magnitude)
		waypointDelta = [2]int{wpLatDelta + waypointDelta[0], wpLongDelta + waypointDelta[1]}
	} else {
		panic(fmt.Errorf("invalid action '%s'", instruction.Action))
	}

	return waypointDelta, latDelta, longDelta
}

func PartTwo(instr string) int {
	inputSlice := parse(instr)

	waypointDelta := [2]int{1, 10}
	lat := 0
	long := 0

	for _, instruction := range inputSlice {
		var (
			lad int
			lod int
		)
		waypointDelta, lad, lod = translateMovementTwo(waypointDelta, instruction)
		lat += lad
		long += lod
	}

	if lat < 0 {
		lat = lat + -2*lat
	}
	if long < 0 {
		long = long + -2 * long
	}

	return lat + long
}
